### PR TITLE
feat(steps): implement target encoding for columns

### PR DESCRIPTION
**Pull request by deepyaman**
# Implementation

Implementation is relatively simple, based on https://github.com/rapidsai/deeplearning/blob/main/RecSys2020Tutorial/solutions/03_3_TargetEncoding.ipynb (look for the "Solution" code).

Assumes numeric targets for now (can be continuous or binary). Going off the assumption right now, if somebody wanted to encode a multiclass target, it would be represented as multiple columns in `target` already. If this is not the case, can do the `OneHotEncode` of target as part of the transform (was my initial plan, until there was some uncertainty around what multiple targets means).

# Examples

**Note that these examples are not meant to be reasonable.** In particular, not sure if multiple targets is the proper representation/meant for a multi-output classification; just using it as random numeric data for now, essentially.

Also, scaling the targets first would probably help a lot...

## Single target

```python
import ibis
import ibisml as ml
from sklearn.pipeline import Pipeline

penguins = ibis.examples.penguins.fetch()
X = penguins.drop("body_mass_g")
y = penguins.body_mass_g

recipe = ml.Recipe(ml.TargetEncode(ml.string()))
pipeline = Pipeline([("recipe", recipe)])
pipeline.fit(X, y)
X_trans = pipeline.transform(X)
```

Result:

```python
     bill_length_mm  bill_depth_mm  flipper_length_mm    year   species    island       sex
0              39.1           18.7              181.0  2007.0  558800.0  189025.0  763675.0
1              39.5           17.4              186.0  2007.0  558800.0  189025.0  637275.0
2              40.3           18.0              195.0  2007.0  558800.0  189025.0  637275.0
3              36.7           19.3              193.0  2007.0  558800.0  189025.0  637275.0
4              39.3           20.6              190.0  2007.0  558800.0  189025.0  763675.0
..              ...            ...                ...     ...       ...       ...       ...
339            44.5           14.3              216.0  2007.0  624350.0  787575.0       NaN
340            46.2           14.4              214.0  2008.0  624350.0  787575.0       NaN
341            47.3           13.8              216.0  2009.0  624350.0  787575.0       NaN
342            44.5           15.7              217.0  2009.0  624350.0  787575.0       NaN
343             NaN            NaN                NaN  2009.0  624350.0  787575.0       NaN

[344 rows x 7 columns]
```

## Multiple targets

```python
import ibis
import ibisml as ml
from sklearn.pipeline import Pipeline

penguins = ibis.examples.penguins.fetch()
# X = penguins.drop("body_mass_g")
X = penguins.drop("flipper_length_mm", "body_mass_g")
# y = penguins.body_mass_g
y = penguins["flipper_length_mm", "body_mass_g"]

recipe = ml.Recipe(ml.TargetEncode(ml.string()))
pipeline = Pipeline([("recipe", recipe)])
pipeline.fit(X, y)
X_trans = pipeline.transform(X)
```

Result:

```python
     bill_length_mm  bill_depth_mm    year  species1  species2  island1   island2     sex1      sex2
0              39.1           18.7  2007.0   28683.0  558800.0   9751.0  189025.0  34357.0  763675.0
1              39.5           17.4  2007.0   28683.0  558800.0   9751.0  189025.0  32565.0  637275.0
2              40.3           18.0  2007.0   28683.0  558800.0   9751.0  189025.0  32565.0  637275.0
3              36.7           19.3  2007.0   28683.0  558800.0   9751.0  189025.0  32565.0  637275.0
4              39.3           20.6  2007.0   28683.0  558800.0   9751.0  189025.0  34357.0  763675.0
..              ...            ...     ...       ...       ...      ...       ...      ...       ...
339            44.5           14.3  2007.0   26714.0  624350.0  35021.0  787575.0      NaN       NaN
340            46.2           14.4  2008.0   26714.0  624350.0  35021.0  787575.0      NaN       NaN
341            47.3           13.8  2009.0   26714.0  624350.0  35021.0  787575.0      NaN       NaN
342            44.5           15.7  2009.0   26714.0  624350.0  35021.0  787575.0      NaN       NaN
343             NaN            NaN  2009.0   26714.0  624350.0  35021.0  787575.0      NaN       NaN

[344 rows x 9 columns]
```

### CODE REVIEW

**Comment by jitingxu1** [comment]
Could we consider handling this? However, I'm inclined to defer to your final decision.

>>> if somebody wanted to encode a multiclass target, it would be represented as multiple columns in target already. If this is not the case, can do the OneHotEncode of target as part of the transform (was my initial plan, until there was some uncertainty around what multiple targets means).

**Comment by deepyaman** [comment]
> Could we consider handling this? However, I'm inclined to defer to your final decision.
> 
> > > > if somebody wanted to encode a multiclass target, it would be represented as multiple columns in target already. If this is not the case, can do the OneHotEncode of target as part of the transform (was my initial plan, until there was some uncertainty around what multiple targets means).

I think I'm not 100% clear on what a `y` with multiple columns represents right now, and want to clarify that first with @jcrist. How this step should be implemented could be different based on whether `y` can be multiple distinct targets, or if `y` is always a single target, potentially encoded.

**Comment by deepyaman** [comment]
> Could we consider handling this? However, I'm inclined to defer to your final decision.
> 
> > > > if somebody wanted to encode a multiclass target, it would be represented as multiple columns in target already. If this is not the case, can do the OneHotEncode of target as part of the transform (was my initial plan, until there was some uncertainty around what multiple targets means).

I think we should consider handling this, but it's not a priority right now IMO. I'll create an issue, but for now we can assume binary or continuous target.

File: ibisml/steps/encode.py [hunk 1]
```diff
@@ -291,3 +291,86 @@ def transform_table(self, table: ir.Table) -> ir.Table:
...
+            joined = table.left_join(
+                encodings, table[c] == encodings[0], lname="left_{name}", rname=""
+            )
+            table = joined.drop(encodings.columns[0], f"left_{c}").rename(
+                {c: encodings.columns[1]}
+                if len(encodings.columns) < 3
+                else {f"{c}{k + 1}": t for k, t in enumerate(encodings.columns[1:])}
+            )
+
+        for k, mean in enumerate(self.target_means_.values()):
```
Comment (jitingxu1) [inline ibisml/steps/encode.py:365]: Replace Null to Mean here does not look very good.  

And Looks use `cases()` could solve the `NULL` issue

Comment (deepyaman) [reply ibisml/steps/encode.py:365]: Using target mean as the encoding for missing columns is consistent with how https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.TargetEncoder.html handles it, I believe. What would you rather replace with?

File: ibisml/steps/encode.py [hunk 2]
```diff
@@ -291,3 +291,86 @@ def transform_table(self, table: ir.Table) -> ir.Table:
...
+                    + self.target_means_[target] * self.smooth
+                ) / (agged[f"{target}_count"] + self.smooth)
+
+            self.encodings_[column] = ibis.memtable(
+                agged.mutate(**target_encodings).drop(target_aggs).to_pyarrow()
+            )
+
+    def transform_table(self, table: ir.Table) -> ir.Table:
+        for c, encodings in self.encodings_.items():
+            joined = table.left_join(
```
Comment (jitingxu1) [inline ibisml/steps/encode.py:356]: Use `cases` could solve null value?

```
┃ sex    ┃ body_mass_g_5160dd ┃
┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━┩
│ string │ float64            │
├────────┼────────────────────┤
│ male   │        4545.684524 │
│ NULL   │        4005.555556 │
│ female │        3862.272727 │
└────────┴────────────────────┘```

something like this

```
map_ = {
    "male": 4545.684524,
    "female":  3862.272727,
    None: 4005.555556 
}

table.mutate(col=col.cases(map_d, map_[None] if None in map_ else 0))

**Comment by jitingxu1** [comment]
I saw we got it merged. @deepyaman 

We may consider encode `None` value as another category, and sklearn did the similar things.

> TargetEncoder considers missing values, such as np.nan or None, as another category and encodes them like any other category.

**Pull request merged**